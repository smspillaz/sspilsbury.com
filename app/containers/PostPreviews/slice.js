import { createSlice } from '@reduxjs/toolkit';

export const initialState = {
  posts: [],
  loading: false,
  error: null,
};

const postPreviewsSlice = createSlice({
  name: 'postPreviews',
  initialState,
  reducers: {
    loadPosts(draft) {
      draft.loading = true;
      draft.error = null;
      draft.posts = [];
    },

    loadPostsSuccess(draft, action) {
      const posts = action.payload;
      draft.loading = false;
      draft.posts = posts;
    },

    loadPostsError(draft, action) {
      draft.error = {
        msg: action.payload.message,
      };
      draft.loading = false;
    },
  },
});

export const {
  loadPosts,
  loadPostsSuccess,
  loadPostsError,
} = postPreviewsSlice.actions;
export const { reducer } = postPreviewsSlice;
