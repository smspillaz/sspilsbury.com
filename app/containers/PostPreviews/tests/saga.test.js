/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { loadPosts, loadPostsSuccess, loadPostsError } from '../slice';

import postsData, { getPosts } from '../saga';

const postsPayload = {
  payload: {
    wordpress: {
      domain: 'smspillaz.wordpress.com',
      count: 2,
    },
  },
};

/* eslint-disable redux-saga/yield-effects */
describe('getPosts Saga', () => {
  let getPostsGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getPostsGenerator = getPosts(postsPayload);

    const callDescriptor = getPostsGenerator.next(postsPayload).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the loadPostsSuccess action if it requests the data successfully', () => {
    const putDescriptor = getPostsGenerator.next(postsPayload).value;
    expect(putDescriptor).toEqual(put(loadPostsSuccess(postsPayload.payload)));
  });

  it('should call the loadPostsError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getPostsGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(loadPostsError(response)));
  });
});

describe('postsDataSaga Saga', () => {
  const postsDataSaga = postsData();

  it('should start task to watch for loadPosts action', () => {
    const takeLatestDescriptor = postsDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(loadPosts.type, getPosts));
  });
});
