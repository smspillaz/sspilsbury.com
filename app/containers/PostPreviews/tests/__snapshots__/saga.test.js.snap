// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPosts Saga should call the loadPostsError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/api/post-previews?wordpressDomain=smspillaz.wordpress.com&wordpressCount=2",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getPosts Saga should dispatch the loadPostsSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/api/post-previews?wordpressDomain=smspillaz.wordpress.com&wordpressCount=2",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
